posts/2023-12-21-async-fn-rpit-in-traits.md

```diff
@@ -0,0 +1,229 @@
+        .map(|p| &p.name)
+}
+```
+
+Starting in Rust 1.75, which hits stable next week, you can use **return-position `impl Trait` in traits** (RPITIT) and trait impls. For example, you could use this to write a trait that returns an iterator:
+
+```rust
+trait Container {
+    fn items(&self) -> impl Iterator<Item = Widget>;
+}
+
+impl Container for MyContainer {
+    fn items(&self) -> impl Iterator<Item = Widget> {
+        self.items.iter().cloned()
+    }
+}
+```
+
+So what does all of this have to do with async functions? Well, async functions are "just sugar" for a function that returns `-> impl Future`. Since that is now permitted in traits, **we also permit you to write traits that use `async fn`**.
+
+```rust
+trait HttpService {
+    async fn fetch(&self, url: Url) -> HtmlBody;
+//  ^^^^^^^^ desugars to:
+//  fn fetch(&self, url: Url) -> impl Future<Output = HtmlBody>;
+}
+```
+
+## Where the gaps lie
+
+### `-> impl Trait` in public traits
+
+The use of `-> impl Trait` is still discouraged for general use in **public** traits, for the reason that users can't use additional bounds on the return type. For example, there is no way to write this function:
+
+```rust!
+fn print_in_reverse(container: impl Container) {
+    for item in container.items().rev() {
+        // ERROR:                 ^^^
+        // the trait `DoubleEndedIterator`
+        // is not implemented for
+        // `impl Iterator<Item = Widget>`
+        eprintln!("{item}");
+    }
+}
+```
+
+In the future we plan to add a solution for this. For now, this feature is best used in internal traits or when you're confident your users won't need additional bounds like this. Otherwise, you should continue using associated types.
+
+### `async fn` in public traits
+
+Since `async fn` desugars to `-> impl Future`, the same limitations apply. If you use bare `async fn` in a public trait today, you'll see a warning. Thankfully, we have a solution that allows using `async fn` in public traits today!
+
+```
+warning: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
+ --> src/lib.rs:7:5
+  |
+7 |     async fn fetch(&self, url: Url) -> HtmlBody;
+  |     ^^^^^
+  |
+help: you can desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`, but these cannot be relaxed without a breaking API change
+  |
+7 -     async fn fetch(&self, url: Url) -> HtmlBody;
+7 +     fn fetch(&self, url: Url) -> impl std::future::Future<Output = HtmlBody> + Send;
+  |
+```
+
+Of particular interest to users of async are `Send` bounds on the returned future. Since users cannot add bounds later, the error message is saying that you as a trait author need to make a choice: Do you want your trait to work with multithreaded, work-stealing executors?
+
+For public traits, we recommend using the `trait-variant` proc macro to let your users choose. Add it to your project with `cargo add trait-variant`, then use it like so:
```

Comment:
```suggestion
For public traits, we recommend using the [`trait-variant`](https://crates.io/crates/trait-variant) proc macro to let your users choose. Add it to your project with `cargo add trait-variant`, then use it like so:
```

Reply:
```suggestion
For public traits, we recommend using the [`trait_variant::make`](https://docs.rs/trait-variant/latest/trait_variant/attr.make.html) proc macro from the [`trait-variant`](https://crates.io/crates/trait-variant) crate to let your users choose. Add it to your project with `cargo add trait-variant`, then use it like so:
```

Reply:
```suggestion
For public traits, we recommend using the [`trait_variant::make`](https://docs.rs/trait-variant/latest/trait_variant/attr.make.html) proc macro to let your users choose. This proc macro is part of the [`trait-variant`](https://crates.io/crates/trait-variant), published by the rust-lang org. Add it to your project with `cargo add trait-variant`, then use it like so:
```

I think we should make clear this is not a "third party" crate somehow.